src/DeterministicNoiseSimulator.cpp

```diff
@@ -400,60 +462,66 @@ void DeterministicNoiseSimulator::ApplyDepolaritationToNode(std::array<qc::Matri
     }
 
-    //e[3] = 0.5*((2-p)*e[3] + p*e[0])
-    if (!old_edges[0].w.approximatelyZero() || !old_edges[3].w.approximatelyZero()) {
-        if (!old_edges[0].w.approximatelyZero() && old_edges[3].w.approximatelyZero()) {
-            e[3].w                = dd->cn.getCached();
-            complex_prob.r->value = probability * 0.5;
-            CN::mul(e[3].w, complex_prob, old_edges[0].w);
-            e[3].p = old_edges[0].p;
-        } else if (old_edges[0].w.approximatelyZero() && !old_edges[3].w.approximatelyZero()) {
+    //e[3] = 0.5*((2-p)*e[3]) + 0.5*(p*e[0])
+    {
+        helper_edge[0].w = dd::Complex::zero;
+        helper_edge[1].w = dd::Complex::zero;
+
+        //helper_edge[0] = 0.5*((2-p)*e[3])
+        if (!e[3].w.approximatelyZero()) {
+            helper_edge[0].w      = dd->cn.getCached();
             complex_prob.r->value = (2 - probability) * 0.5;
-            CN::mul(e[3].w, complex_prob, old_edges[3].w);
-            e[3].p = old_edges[3].p;
-        } else {
+            CN::mul(helper_edge[0].w, complex_prob, e[3].w);
+            helper_edge[0].p = e[3].p;
+        }
+
+        //helper_edge[1] = 0.5*p*e[0]
+        if (!old_e0_edge.w.approximatelyZero()) {
+            helper_edge[1].w      = dd->cn.getCached();
             complex_prob.r->value = probability * 0.5;
-            CN::mul(helper_edge[0].w, complex_prob, old_edges[0].w);
-            complex_prob.r->value = (2 - probability) * 0.5;
-            CN::mul(helper_edge[1].w, complex_prob, old_edges[3].w);
-            helper_edge[0].p = old_edges[0].p;
-            helper_edge[1].p = old_edges[3].p;
+            CN::mul(helper_edge[1].w, complex_prob, old_e0_edge.w);
+            helper_edge[1].p = old_e0_edge.p;
+        }
+
+        //e[3] = helper_edge[0] + helper_edge[1]
+        if (e[3].w != dd::Complex::zero) {
             dd->cn.returnToCache(e[3].w);
-            e[3] = dd->add2(helper_edge[0], helper_edge[1]);
         }
-    }
-    for (auto& old_edge: old_edges) {
-        if (!old_edge.w.approximatelyZero()) {
-            dd->cn.returnToCache(old_edge.w);
+        e[3] = dd->add2(helper_edge[0], helper_edge[1]);
+
+        if (helper_edge[0].w != dd::Complex::zero) {
+            dd->cn.returnToCache(helper_edge[0].w);
+        }
+        if (helper_edge[1].w != dd::Complex::zero) {
+            dd->cn.returnToCache(helper_edge[1].w);
         }
     }
-    //    helper_edge[0].w.r->next->next->next->next->next->next = ComplexCache_Avail;
-    dd->cn.returnToCache(helper_edge[0].w);
-    dd->cn.returnToCache(helper_edge[1].w);
+    dd->cn.returnToCache(old_e0_edge.w);
     dd->cn.returnToCache(complex_prob);
 }
 
-std::map<std::string, double> DeterministicNoiseSimulator::AnalyseState(dd::QubitCount nr_qubits, bool full_state) const {
-    std::map<std::string, double> measure_result = {};
-
-    double p0, p1;
+std::map<std::string, double> DeterministicNoiseSimulator::AnalyseState(dd::QubitCount nr_qubits, bool full_state = false) {
```

Comment:
this should also reside in the DD Package as does the `measure` routine

Reply:
Done. I moved it there and also renamed it.